tensorflow_quantum/core/ops/tfq_adj_grad_op.cc

```diff
@@ -0,0 +1,403 @@
+    using StateSpace = Simulator::StateSpace;
+    using State = StateSpace::State;
+
+    auto DoWork = [&](int start, int end) {
+      // Begin simulation.
+      int largest_nq = 1;
+      State sv = StateSpace(largest_nq, tfq_for).CreateState();
+      State scratch = StateSpace(largest_nq, tfq_for).CreateState();
+      State scratch2 = StateSpace(largest_nq, tfq_for).CreateState();
+
+      for (int i = start; i < end; i++) {
+        int nq = num_qubits[i];
+        Simulator sim = Simulator(nq, tfq_for);
+        StateSpace ss = StateSpace(nq, tfq_for);
+        if (nq > largest_nq) {
+          // need to switch to larger statespace.
+          largest_nq = nq;
+          sv = ss.CreateState();
+          scratch = ss.CreateState();
+          scratch2 = ss.CreateState();
+        }
+
+        // (#679) Just ignore empty program
+        if (qsim_circuits[i].gates.size() == 0) {
+          continue;
+        }
+
+        ss.SetStateZero(sv);
+        for (int j = 0; j < partial_fused_circuits[i].size(); j++) {
+          for (int k = 0; k < partial_fused_circuits[i][j].size(); k++) {
+            qsim::ApplyFusedGate(sim, partial_fused_circuits[i][j][k], sv);
+          }
+          if (j == partial_fused_circuits[i].size() - 1) {
+            break;
+          }
+          qsim::ApplyGate(
+              sim, qsim_circuits[i].gates[gradient_gates[i][j].index], sv);
+        }
+
+        // sv now contains psi
+        // scratch contains (sum_i paulis_sums[i] * downstream_grads[i])|psi>
+        // scratch2 now contains psi as well.
+        AccumulateOperators(pauli_sums[i], downstream_grads[i], sim, ss, sv,
+                            scratch2, scratch);
+
+        for (int j = partial_fused_circuits[i].size() - 1; j >= 0; j--) {
+          for (int k = partial_fused_circuits[i][j].size() - 1; k >= 0; k--) {
+            ApplyFusedGateDagger(sim, partial_fused_circuits[i][j][k], sv);
+            ApplyFusedGateDagger(sim, partial_fused_circuits[i][j][k], scratch);
+          }
+          if (j == 0) {
+            // last layer will have no parametrized gates so can break.
+            break;
+          }
+
+          // Hit a parameterized gate.
+          ApplyGateDagger(
+              sim, qsim_circuits[i].gates[gradient_gates[i][j - 1].index], sv);
+          for (int k = 0; k < gradient_gates[i][j - 1].grad_gates.size(); k++) {
+            // Copy sv onto scratch2 in anticipation of non-unitary "gradient
+            // gate".
+            ss.CopyState(sv, scratch2);
+            qsim::ApplyGate(sim, gradient_gates[i][j - 1].grad_gates[k], sv);
+
+            // don't need not-found check since this is done upstream already.
+            const auto it = maps[i].find(gradient_gates[i][j - 1].params[k]);
+            const int loc = it->second.first;
+            (*output_tensor)(i, loc) += ss.RealInnerProduct(sv, scratch) +
+                                        ss.RealInnerProduct(scratch, sv);
```

Comment:
I think we need some comments like the following for the future (maybe) analytic gradient methods:
```
// Apply finite differencing for adjoint gradients.
// Finite differencing enables applying multiple `gradient_gate` of a symbol at the same circuit.
// For analytic methods like parameter-shift we need to apply a single `gradient_gate`
// per a symbol.
```

Reply:
OR if we don't have any plan to use parameter-shift for adjoint gradient, please ignore this :)

Reply:
This could definitely help others reading over it. Added.